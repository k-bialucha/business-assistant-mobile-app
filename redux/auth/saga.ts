import * as Facebook from 'expo-facebook';
import decode from 'jwt-decode';
import { FACEBOOK_APP_ID } from 'react-native-dotenv';
import { call, delay, put, takeLatest } from 'redux-saga/effects';

import { loginUser, signupUser } from '../../utils/apiCalls/authorization';
import firebase, { myFirebaseApp } from '../../utils/firebase';

import {
  loginFailure,
  loginSuccess,
  signupFailure,
  signupSuccess,
} from './actions';
import {
  LOGIN,
  LOGIN_WITH_FACEBOOK,
  LoginAction,
  SIGNUP,
  SignupAction,
} from './types';

export function* loginSaga({ payload: { email, password } }: LoginAction) {
  try {
    yield delay(1000);

    const response = yield call(loginUser, email, password);

    const { idToken: token, localId: userId } = response;

    yield put(loginSuccess(token, { name: email, id: userId }));
  } catch (error) {
    if (error instanceof Error) {
      yield put(loginFailure(error.message));
    } else {
      yield put(loginFailure('Login Error'));
    }
  }
}

export function* signupSaga({
  payload: { email, password, phone },
}: SignupAction) {
  try {
    yield delay(1000);

    const response = yield call(signupUser, email, password, phone);

    const { idToken: token, localId: userId } = response;

    yield put(signupSuccess(token, { name: email, id: userId }));
  } catch (error) {
    if (error instanceof Error) {
      yield put(signupFailure(error.message));
    } else {
      yield put(signupFailure('Signup Error'));
    }
  }
}

export function* loginWithFacebookSaga() {
  yield call(
    Facebook.initializeAsync,
    FACEBOOK_APP_ID,
    'Personal Business Assistant'
  );
  try {
    const { type, token } = yield call(Facebook.logInWithReadPermissionsAsync, {
      permissions: ['public_profile', 'email'],
    });

    if (type === 'success') {
      // https://stackoverflow.com/questions/53678410/jest-test-the-current-environment-does-not-support-the-specified-persistence-t
      // look like without it login works the same, to delete?

      // const persistence =
      //   process.env.NODE_ENV === 'test'
      //     ? firebase.auth.Auth.Persistence.NONE
      //     : firebase.auth.Auth.Persistence.LOCAL;

      // yield call(
      //   [firebase.auth(), firebase.auth().setPersistence],
      //   firebase.auth.Auth.Persistence.LOCAL
      // );

      const credential = yield call(
        firebase.auth.FacebookAuthProvider.credential,
        token
      );

      // way to call without redux-saga-firebase
      // const { user } = yield call(
      //   [firebase.auth(), firebase.auth().signInWithCredential],
      //   credential
      // );

      const { user } = yield call(
        myFirebaseApp.auth.signInWithCredential,
        credential
      );

      const jwtToken = yield user.getIdToken();
      const { user_id: userId, name, picture } = decode(jwtToken);

      yield put(loginSuccess(token, { name, id: userId, image: picture }));
    } else if (type === 'cancel') {
      throw new Error('Login to facebook canceled');
    }
  } catch (error) {
    if (error instanceof Error) {
      yield put(loginFailure(error.message));
    } else {
      yield put(loginFailure('Login Error'));
    }
  }
}

export default function* watchSaga() {
  yield takeLatest(LOGIN, loginSaga);
  yield takeLatest(SIGNUP, signupSaga);
  yield takeLatest(LOGIN_WITH_FACEBOOK, loginWithFacebookSaga);
}
