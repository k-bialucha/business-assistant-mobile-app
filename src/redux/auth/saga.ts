import { Alert } from 'react-native';

import * as Facebook from 'expo-facebook';
import * as Google from 'expo-google-app-auth';
import { call, delay, put, takeLatest } from 'redux-saga/effects';

import { ANDROID_CLIENT_ID, FACEBOOK_APP_ID, IOS_CLIENT_ID } from '~env';

import { firebase, getCurrentUser } from '~/utils/firebase';

import {
  loginFailure,
  loginSuccess,
  signupFailure,
  signupSuccess,
} from './actions';
import {
  LOGIN,
  LOGIN_WITH_FACEBOOK,
  LOGIN_WITH_GOOGLE,
  LoginAction,
  LOGOUT,
  RESET_PASSWORD,
  ResetPasswordAction,
  SIGNUP,
  SignupAction,
  TRY_AUTO_LOGIN,
  UserData,
} from './types';

export function* loginSaga({ payload: { email, password } }: LoginAction) {
  try {
    yield delay(1000);

    const { user } = yield call(
      [firebase.auth(), firebase.auth().signInWithEmailAndPassword],
      email,
      password
    );

    yield put(
      loginSuccess({
        username: user.displayName || email,
        id: user.uid,
        image: user.photoURL,
      })
    );
  } catch (error) {
    if (error instanceof Error) {
      yield put(loginFailure(error.message));
    } else {
      yield put(loginFailure('Login Error'));
    }
  }
}

export function* signupSaga({ payload: { email, password } }: SignupAction) {
  try {
    yield delay(1000);

    const { user } = yield call(
      [firebase.auth(), firebase.auth().createUserWithEmailAndPassword],
      email,
      password
    );
    // TODO: save phone number for registered user

    yield put(signupSuccess({ username: user.email, id: user.uid }));
  } catch (error) {
    if (error instanceof Error) {
      yield put(signupFailure(error.message));
    } else {
      yield put(signupFailure('Signup Error'));
    }
  }
}

// FIXME: still log in with  Fb in iOS by Facebook App doesn't work (not redirecting back to app)
export function* loginWithFacebookSaga() {
  yield call(
    Facebook.initializeAsync,
    FACEBOOK_APP_ID,
    'Personal Business Assistant'
  );

  try {
    const response: Facebook.FacebookLoginResult = yield call(
      Facebook.logInWithReadPermissionsAsync,
      {
        permissions: ['public_profile', 'email'],
      }
    );

    if (response.type === 'success') {
      const { token } = response;

      const credential = yield call(
        firebase.auth.FacebookAuthProvider.credential,
        token
      );

      const { user } = yield call(
        [firebase.auth(), firebase.auth().signInWithCredential],
        credential
      );

      yield put(
        loginSuccess({
          username: user.displayName,
          id: user.uid,
          image: user.photoURL,
        })
      );
    } else if (response.type === 'cancel') {
      throw new Error('Login to facebook canceled');
    }
  } catch (error) {
    if (error instanceof Error) {
      yield put(loginFailure(error.message));
    } else {
      yield put(loginFailure('Login Error'));
    }
  }
}

export function* loginWithGoogleSaga() {
  try {
    const { type, idToken, accessToken } = yield call(Google.logInAsync, {
      iosClientId: IOS_CLIENT_ID,
      androidClientId: ANDROID_CLIENT_ID,
      scopes: ['profile', 'email'],
    });

    if (type === 'success') {
      const credential = yield call(
        firebase.auth.GoogleAuthProvider.credential,
        idToken,
        accessToken
      );

      const { user } = yield call(
        [firebase.auth(), firebase.auth().signInWithCredential],
        credential
      );

      yield put(
        loginSuccess({
          username: user.displayName,
          id: user.uid,
          image: user.photoURL,
        })
      );
    } else if (type === 'cancel') {
      throw new Error('Login to google canceled');
    }
  } catch (error) {
    if (error instanceof Error) {
      yield put(loginFailure(error.message));
    } else {
      yield put(loginFailure('Login Error'));
    }
  }
}

export function* resetPasswordSaga({
  payload: { email },
}: ResetPasswordAction) {
  try {
    yield call(
      [firebase.auth(), firebase.auth().sendPasswordResetEmail],
      email
    );

    Alert.alert(
      'Reset Password',
      'Link to reset password has been sent to the email'
    );
  } catch (error) {
    if (error instanceof Error) {
      Alert.alert('Reset Password', error.message);
    } else {
      Alert.alert('Reset Password', 'Something went wrong');
    }
  }
}

export function* tryAutoLoginSaga() {
  try {
    const user = yield call(getCurrentUser);

    if (user) {
      // we can access the token by calling method below
      // const token = yield user.getIdToken();

      const userData: UserData = {
        id: user.uid,
        username: user.displayName || user.email || `User ID ${user.uid}`,
      };

      if (user.photoURL) userData.image = user.photoURL;

      yield put(loginSuccess(userData));
    } else {
      throw new Error();
    }
  } catch (error) {
    yield put(loginFailure());
  }
}

export function* logoutSaga() {
  yield call([firebase.auth(), firebase.auth().signOut]);
}

export default function* watchSaga() {
  yield takeLatest(LOGIN, loginSaga);
  yield takeLatest(SIGNUP, signupSaga);
  yield takeLatest(LOGIN_WITH_FACEBOOK, loginWithFacebookSaga);
  yield takeLatest(LOGIN_WITH_GOOGLE, loginWithGoogleSaga);
  yield takeLatest(RESET_PASSWORD, resetPasswordSaga);
  yield takeLatest(TRY_AUTO_LOGIN, tryAutoLoginSaga);
  yield takeLatest(LOGOUT, logoutSaga);
}
