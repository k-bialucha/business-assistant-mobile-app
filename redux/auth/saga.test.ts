import { put } from 'redux-saga/effects';

import { loginSaga, signupSaga } from './saga';
import {
  DOMAIN_NAME,
  LOGIN,
  LOGIN_FAILURE,
  LOGIN_SUCCESS,
  SIGNUP,
  SIGNUP_FAILURE,
  SIGNUP_SUCCESS,
} from './types';

describe(`${DOMAIN_NAME}/saga`, () => {
  describe('loginSaga', () => {
    it('handles successful login', () => {
      const someApiResponse = {
        idToken: 'token-123',
        localId: '12345',
      };
      const generator = loginSaga({
        type: LOGIN,
        payload: { username: 'some-username', password: 'strong-password' },
      });

      const delayDescriptor = generator.next().value;
      const apiCallDescriptor = generator.next().value;
      const putDescriptor = generator.next(someApiResponse).value;

      expect(delayDescriptor).toMatchSnapshot();
      expect(apiCallDescriptor).toMatchSnapshot();
      expect(putDescriptor).toEqual(
        put({
          type: LOGIN_SUCCESS,
          payload: { token: expect.any(String), userId: expect.any(String) },
        })
      );
      expect(generator.next().done).toBe(true);
    });

    it('handles unsuccessful login', () => {
      const someError = new Error('Bad Credentials');
      const generator = loginSaga({
        type: LOGIN,
        payload: { username: 'some-username', password: 'badpass' },
      });

      const delayDescriptor = generator.next().value;
      const apiCallDescriptor = generator.next().value;
      const putDescriptor = generator.throw(someError).value;

      expect(delayDescriptor).toMatchSnapshot();
      expect(apiCallDescriptor).toMatchSnapshot();
      expect(putDescriptor).toEqual(
        put({ type: LOGIN_FAILURE, payload: someError.message })
      );
      expect(generator.next().done).toBe(true);
    });
  });

  describe('signupSaga', () => {
    it('handles successful signup', () => {
      const someApiResponse = {
        idToken: 'token-123',
        localId: '12345',
      };
      const generator = signupSaga({
        type: SIGNUP,
        payload: {
          email: 'some-email@example.com',
          password: 'strong-password',
        },
      });

      const delayDescriptor = generator.next().value;
      const apiCallDescriptor = generator.next().value;
      const putDescriptor = generator.next(someApiResponse).value;

      expect(delayDescriptor).toMatchSnapshot();
      expect(apiCallDescriptor).toMatchSnapshot();
      expect(putDescriptor).toEqual(
        put({
          type: SIGNUP_SUCCESS,
          payload: { token: expect.any(String), userId: expect.any(String) },
        })
      );
      expect(generator.next().done).toBe(true);
    });

    it('handles unsuccessful signup', () => {
      const someError = new Error('Bad Credentials');
      const generator = signupSaga({
        type: SIGNUP,
        payload: { email: 'some-email@example.com', password: 'badpass' },
      });

      const delayDescriptor = generator.next().value;
      const apiCallDescriptor = generator.next().value;
      const putDescriptor = generator.throw(someError).value;

      expect(delayDescriptor).toMatchSnapshot();
      expect(apiCallDescriptor).toMatchSnapshot();
      expect(putDescriptor).toEqual(
        put({ type: SIGNUP_FAILURE, payload: someError.message })
      );
      expect(generator.next().done).toBe(true);
    });
  });
});
