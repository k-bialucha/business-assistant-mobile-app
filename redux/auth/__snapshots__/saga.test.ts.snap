// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth/saga loginSaga handles successful login 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      1000,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginSaga handles successful login 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "some-email",
      "strong-password",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginSaga handles successful login 3`] = `
Promise {
  "_40": 0,
  "_55": null,
  "_65": 0,
  "_72": null,
}
`;

exports[`auth/saga loginSaga handles unsuccessful login 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      1000,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginSaga handles unsuccessful login 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "some-email",
      "badpass",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithFacebookSaga handles successful signup through facebook 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "facebook-app-id",
      "Personal Business Assistant",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithFacebookSaga handles successful signup through facebook 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "permissions": Array [
          "public_profile",
          "email",
        ],
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithFacebookSaga handles successful signup through facebook 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "mocked-token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithFacebookSaga handles successful signup through facebook 4`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "token": "",
        "type": "",
        "user": "user-object",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithFacebookSaga handles successful signup through facebook 5`] = `
Promise {
  "_40": 0,
  "_55": null,
  "_65": 0,
  "_72": null,
}
`;

exports[`auth/saga loginWithGoogleSaga handles successful signup through google 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "androidClientId": undefined,
        "iosClientId": undefined,
        "scopes": Array [
          "profile",
          "email",
        ],
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithGoogleSaga handles successful signup through google 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "mocked-id-token",
      "mocked-token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithGoogleSaga handles successful signup through google 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "accessToken": "mocked-token",
        "idToken": "mocked-id-token",
        "type": "",
        "user": "user-object",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga loginWithGoogleSaga handles successful signup through google 4`] = `
Promise {
  "_40": 0,
  "_55": null,
  "_65": 0,
  "_72": null,
}
`;

exports[`auth/saga signupSaga handles successful signup 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      1000,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga signupSaga handles successful signup 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "some-email@example.com",
      "strong-password",
      undefined,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga signupSaga handles successful signup 3`] = `
Promise {
  "_40": 0,
  "_55": null,
  "_65": 0,
  "_72": null,
}
`;

exports[`auth/saga signupSaga handles unsuccessful signup 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      1000,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`auth/saga signupSaga handles unsuccessful signup 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "some-email@example.com",
      "badpass",
      undefined,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
